### PR TITLE
Np 44571 literary arts audio visual subtype other

### PR DESCRIPTION
**Pull request by brinxmat**
- Allows description of `Other` types
- You can look at the comments
- Migrates old data to new format
- Use the file-filter functionality of GitHub to remove e.g. JSON and YAML as these are mostly uninteresting (though. see comment on `LiteraryArts.json`

### CODE REVIEW

File: nva-datamodel-java/src/test/java/no/unit/nva/LiteraryArtsAudioVisualMigrationTest.java [hunk 1]
```diff
@@ -75,6 +77,47 @@ private static String nullIsbns() {
                 + "        }";
     }
 
+    private static String otherWithDescription() {
```
Comment (brinxmat) [inline nva-datamodel-java/src/test/java/no/unit/nva/LiteraryArtsAudioVisualMigrationTest.java:80]: Added two new tests for the possible new variants. The old variants should be migrated.

File: nva-datamodel-java/src/main/java/no/unit/nva/model/instancetypes/artistic/literaryarts/manifestation/LiteraryArtsAudioVisualSubtype.java [hunk 2]
```diff
@@ -1,43 +1,81 @@
...
+    public static LiteraryArtsAudioVisualSubtype create(Object subtype) {
+        if (subtype instanceof String) {
+            return fromJson(LiteraryArtsAudioVisualSubtypeEnum.parse((String) subtype), null);
+        } else if (subtype instanceof LiteraryArtsAudioVisualSubtypeEnum) {
+            return fromJson((LiteraryArtsAudioVisualSubtypeEnum) subtype, null);
+        } else if (subtype instanceof Map) {
+            return parseCurrentImplementation((LinkedHashMap<String, String>) subtype);
+        } else {
+            throw new IllegalArgumentException("Invalid subtype");
+        }
```
Comment (brinxmat) [inline nva-datamodel-java/src/main/java/no/unit/nva/model/instancetypes/artistic/literaryarts/manifestation/LiteraryArtsAudioVisualSubtype.java:33]: The code here allows us to migrate Strings to things

File: nva-datamodel-java/src/test/java/no/unit/nva/model/instancetypes/artistic/literaryarts/manifestation/LiteraryArtsManifestationMigrationTest.java [hunk 3]
```diff
@@ -23,8 +23,12 @@ void shouldCreateLiteraryArtsMonographWhenIsbnIsString() {
...
     @Test
     void shouldCreateLiteraryArtsAudioVisualWhenIsbnIsString() {
-        var literaryArtsAudioVisual = new LiteraryArtsAudioVisual(LiteraryArtsAudioVisualSubtype.PODCAST,
+        var literaryArtsAudioVisual = new LiteraryArtsAudioVisual(randomLiteraryArtsAudioVisualSubtype(),
                 randomPublishingHouse(), randomPublicationDate(), randomIsbn13(), randomInteger());
         assertThat(literaryArtsAudioVisual.getIsbnList(), hasSize(1));
     }
+
+    private LiteraryArtsAudioVisualSubtype randomLiteraryArtsAudioVisualSubtype() {
+        return new LiteraryArtsAudioVisualSubtype(LiteraryArtsAudioVisualSubtypeEnum.PODCAST);
```
Comment (monaullah) [inline nva-datamodel-java/src/test/java/no/unit/nva/model/instancetypes/artistic/literaryarts/manifestation/LiteraryArtsManifestationMigrationTest.java:32]: Podcast is not a random subtype?

Comment (monaullah) [reply nva-datamodel-java/src/test/java/no/unit/nva/model/instancetypes/artistic/literaryarts/manifestation/LiteraryArtsManifestationMigrationTest.java:32]: This code (from PublicationInstanceBuilder.java) looks more correct:
```
private static LiteraryArtsAudioVisualSubtype randomLiteraryArtsAudioVisualSubtype() {
        var subtype = randomElement(LiteraryArtsAudioVisualSubtypeEnum.values());

        return LiteraryArtsAudioVisualSubtypeEnum.OTHER.equals(subtype)
                ? LiteraryArtsAudioVisualSubtype.createOther(randomString())
                : LiteraryArtsAudioVisualSubtype.create(subtype);
    }
```

**Review by monaullah — APPROVED** [review]

File: nva-datamodel-java/src/main/java/no/unit/nva/model/instancetypes/artistic/literaryarts/manifestation/LiteraryArtsAudioVisualSubtypeOther.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
+package no.unit.nva.model.instancetypes.artistic.literaryarts.manifestation;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import nva.commons.core.JacocoGenerated;
+
+import java.util.Objects;
+
+public class LiteraryArtsAudioVisualSubtypeOther extends LiteraryArtsAudioVisualSubtype {
```
Comment (StigNorland) [inline nva-datamodel-java/src/main/java/no/unit/nva/model/instancetypes/artistic/literaryarts/manifestation/LiteraryArtsAudioVisualSubtypeOther.java:10]: That `NON_NULL` forces you to do this, I will just look another way....

**Review by anetteOlli — DISMISSED** [review]

File: documentation/ChapterConferenceAbstract.json [hunk 5]
```diff
@@ -1,171 +1,171 @@
...
-    "description" : "9xH4DZxOVh167Gz3",
+    "npiSubjectHeading" : "pGFppN8dbjVInv",
+    "tags" : [ "1hw1xFxyeNP4HMP" ],
+    "description" : "mBXeGm8TP1kR5EEQuNx",
     "reference" : {
       "type" : "Reference",
       "publicationContext" : {
         "type" : "Anthology",
-        "id" : "https://www.example.com/sTFMfkd77s"
+        "id" : "https://www.example.com/hFAwx5nEWZEb1S2M4M"
```
[the remark this answers is not in the document]
Comment (brinxmat) [reply documentation/ChapterConferenceAbstract.json:91]: Hmm, possible old data generation…let's see

File: nva-datamodel-testutils/src/main/java/no/unit/nva/model/testing/PublicationInstanceBuilder.java [hunk 6]
```diff
@@ -775,6 +776,14 @@ private static List<LiteraryArtsManifestation> randomLiteraryArtsManifestationLi
         return List.of(monograph, audioVisual, performance, web);
     }
 
+    public static LiteraryArtsAudioVisualSubtype randomLiteraryArtsAudioVisualSubtype() {
+        var subtype = randomElement(LiteraryArtsAudioVisualSubtypeEnum.values());
```
Comment (StigNorland) [inline nva-datamodel-testutils/src/main/java/no/unit/nva/model/testing/PublicationInstanceBuilder.java:780]: 👍

**Review by StigNorland — APPROVED** [review]

**Pull request merged**